SpamAssassin/clamav/clamav-0.88.2/docs/man/freshclam.1
.\" Manual page created by Tomasz Kojm, 20020415
.TH "freshclam" "1" "May 11, 2005" "Tomasz Kojm" "Clam AntiVirus"
.SH "NAME"
.LP 
freshclam \- update virus databases
.SH "SYNOPSIS"
.LP 
freshclam [options]
.SH "DESCRIPTION"
.LP 
freshclam updates the virus database. It's a part of the Clam AntiVirus package. It requires an Internet connection.
.SH "OPTIONS"
.LP 
Freshclam reads its configuration from freshclam.conf. The settings can be overwritten with command line options.
.TP 
\fB\-h, \-\-help\fR
Output help information and exit.
.TP 
\fB\-V, \-\-version\fR
Print version number and exit.
.TP 
\fB\-v, \-\-verbose\fR
Be verbose. This option causes freshclam to print many additional informations.
.TP 
\fB\-\-quiet\fR
Be quiet \- output only error messages.
.TP 
\fB\-\-stdout\fR
Write all messages to the standard output (stdout), instead of the standard error output (stderr).
.TP 
\fB\-l FILE, \-\-log=FILE\fR
Save download report in FILE.
.TP 
\fB\-\-datadir=DIRECTORY\fR
Install new database in DIRECTORY. The directory must be writable for the 'clamav' user or unprivileged user running freshclam.
.TP 
\fB\-u USER, \-\-user USER\fR
Run as USER. By default (when started by root) freshclam drops privileges and works as the 'clamav' user.
.TP 
\fB\-d, \-\-daemon\fR
Run in a daemon mode. This option requires \-\-checks.
\fB\-p FILE, \-\-pid=FILE\fR
Save daemon's pid in FILE.
.TP 
\fB\-\-no\-dns\fR
This option forces old non\-DNS verification method (without a TTL delay).
.TP 
\fB\-c #n, \-\-checks=#n\fR
Check #n times per day for a new database. #n must be between 1 and 50.
.TP 
\fB\-\-daemon\-notify=/path/to/clamd.conf\fR
Notify the daemon about the new database. By default it reads a hardcoded config file but you can use an another one. Both local and TCP sockets are supported.
.TP 
\fB\-a IP, \-\-local\-address=IP\fR
Use (local) IP for HTTP downloads. Useful for multi\-homed systems. If binding fails for whatever reason, a warning is issued and freshclam behaves like without this flag.
.TP 
\fB\-\-on\-error\-execute=COMMAND\fR
Execute COMMAND if error occurred. Remember, that virus database freshness is the most important thing in anti\-virus system. With this option freshclam can alert you (eg. send SMS) when something is going wrong.
.TP 
\fB\-\-on\-update\-execute=COMMAND\fR
Execute COMMAND after successful update.
.TP 
\fB\-\-on\-outdated\-execute=COMMAND\fR
Execute COMMAND when freshclam reports outdated version. In the command string %v will be replaced by the new version number.
.SH "EXAMPLES"
.LP 
.TP 
(0) Download database to default directory:

\fBfreshclam\fR
.TP 
(1) Download database to current directory:

\fBfreshclam \-\-datadir=.\fR
.TP 
(2) Run as a daemon and check 2 times per day for new database:

\fBfreshclam \-d \-c 2\fR
.SH "RETURN CODES"
0 : Database successfully updated.
.TP 
1 : Database is up\-to\-date.
.TP 
40: Unknown option passed.
.TP 
50: Can't change directory.
.TP 
51: Can't check MD5 sum.
.TP 
52: Connection (network) problem.
.TP 
53: Can't unlink file.
.TP 
54: MD5 or digital signature verification error.
.TP 
55: Error reading file.
.TP 
56: Config file error.
.TP 
57: Can't create new file.
.TP 
58: Can't read database from remote server.
.TP 
59: Mirrors are not fully synchronized (try again later).
.TP 
60: Can't get information about clamav user from /etc/passwd.
.TP 
61: Can't drop privileges.
.TP 
62: Can't initialize logger.
.SH "FILES"
.LP 
/usr/local/etc/freshclam.conf
.SH "CREDITS"
Please check the full documentation for credits.
.SH "AUTHOR"
.LP 
Tomasz Kojm <tkojm@clamav.net>
.SH "SEE ALSO"
.LP 
clamd(8), clamdscan(1), clamscan(1), sigtool(1), clamd.conf(5)


gdb/ChangeLog
2006-05-02  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Default to building with -gdwarf-2.
	(install-gdb-macosx, install-gdb-fat): Create dSYM files,
	strip the binaries that are left in SYMROOT.

2006-05-01  Jason Molenda  (jmolenda@apple.com)

	* gdb.sh: Unset the DYLD_* env vars before invoking gdb;
	dyld-60 and later issue a warning when a setgid program is invoked
	and these env vars are not honored by dyld.  
	* Makefile (check): Properly quote filename check so that multiple
	matches don't yield a shell error.

2006-04-18  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Don't remove the symroot between phases.

2006-04-17  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Make installed MacsBug plugin setgid procmod.

2006-04-17  Jason Molenda  (jmolenda@apple.com)

	* gdb.sh: If the parent process is "CodeWarrior", invoke
	gdb-for-codewarrior.

2006-01-31  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Remove a bunch of unused sections of the Makefile.

2006-01-24  Stan Shebs  <shebs@apple.com>

	* Makefile (GDB_VERSION): Update.

2005-12-16  Jason Molenda  (jmolenda@apple.com)

	* Makefile (CDEBUGFLAGS): Drop -mdynamic-no-pic for now; libintl
	builds in a way that requires the relocations to be present even
	though only a .a file is finally generated.

2005-12-16  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Remove references to mmalloc.

2005-12-16  Jim Ingham  <jingham@apple.com>

	* Makefile: Pass the GDB_RC_VERSION separately to the gdb 
	submake.
	
2005-12-01  Ronnie Misra (rgmisra@apple.com)

        * gdb.sh: Update to recognize new translate program path/command 
        line parameter.

2005-11-09  Jason Molenda  (jmolenda@apple.com)

	* gdb.sh (oah750_binary): Add "-execOAH" command line parameter
	when invoking oah750 so it knows to leave our procmod setgid 
	permissions in tact.

2005-10-19  Jason Molenda  (jmolenda@apple.com)

        * gdb.sh: Copy DYLD_* into GDB_DYLD_* to get past the setgid barrier.

2005-09-30  Jason Molenda  (jmolenda@apple.com)

        * Makefile: make the gdb binaries setgid procmod.

2005-06-07  Jason Molenda  (jmolenda@apple.com)

	* Makefile (GDB_RC_VERSION): Bump to 450.  Tumeric will be
	in the 420-449 range.

2005-05-28  Jason Molenda  (jmolenda@apple.com)

	* gdb.sh: Update --help text.

2005-05-18  Jason Molenda  (jmolenda@apple.com)

	* cache-symfiles.sh: Remove.
	* Makefile: Don't install cache-symfiles.sh.

2005-05-03  Jason Molenda  (jmolenda@apple.com)

	* gdb.sh: Add --oah750 command line option.

2005-04-14  Jason Molenda  (jmolenda@apple.com)

	* gdb.sh: Add a -arch command line option.

2005-01-18  Jason Molenda  (jmolenda@apple.com)

	* Makefile (GDB_VERSION): Update to the last merge.

2004-12-06  Jason Molenda  (jmolenda@apple.com)

	* Info-template.plist: It looks like CFBundleDevelopmentRegion needs
	to be defined; whatever.  Looking at two installed frameworks I see 
	English is an acceptable setting, so there we go.  These are just 
	private framework header files used for building gdb plug-ins anyway.

2004-08-09  Klee Dienes  <kdienes@apple.com>

	* Makefile: Post-process the man page to refer to
	$(SYSTEM_DEVELOPER_TOOLS_DOC_DIR) instead of hard-coding (see
	change to gdb/gdb.1).

2004-06-16  Jason Molenda  (jmolenda@apple.com)

	* ntbuild.sh, ntconfig.sh: Removed.
	* dpkg/control: Removed.
	* ChangeLog.binutils: Removed.

2004-06-16  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Don't install binutils, unnecessary frameworks.
	Don't create /usr/lib, /usr/libexec/binutils directories.

2004-06-11  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Quote documentation directory pathnames in case
	a space character is added.

2004-06-08  Jim Ingham  <jingham@apple.com>

        * cache-symfiles.sh: For now, don't emit cached symfiles
        if the original library has any stabs.  gdb crashes when
        these are present (Radar 3418798)
        Also deleted the setting of "name" in the second (library)
        search, since we don't actually use it.
	Also, run the script far enough to delete the old symfile
	directory, since then gdb won't bother to look at a bunch
	of symfiles which it surely can't use.
	
2004-06-08  Jason Molenda  (jmolenda@apple.com)

	* Makefile (SYSTEM_DEVELOPER_TOOLS_DOC_DIR): Set from
	pb_makefiles' plaform-variables.make if possible. 

2004-06-07  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Move GDB_RC_VERSION up where I can see it.
	(SYSTEM_DEVELOPER_TOOLS_DOC_DIR): New variable to hold the location
	of the installed documentation, cf r. 3678942.
	(install-binutils-common): Don't install binutils docs.
	(install-binutils-macosx): Don't install binutils docs.
	(install-chmod-macosx): Update chmod call to avoid deprecation warning.

2004-05-17  Jason Molenda  (jmolenda@apple.com)

	* Makefile: Bump GDB_RC_VERSION to new range.

2004-05-13  Jason Molenda  (jmolenda@apple.com)

	* cached-symfiles.sh: Suppress cacheing of symfiles until Klee
	has a chance to make it work again in the newly merged sources.

2003-10-03  Jason Molenda  (jmolenda@apple.com)

        * cache-symfiles.sh: Run nm on dylibs to ensure they're well formed
        before feeding them to gdb.

2003-07-24  Klee Dienes <kdienes@apple.com>

	* cache-symfiles.sh: Use the proper prefix for dyld; no longer
	remove the dyld.syms from the cached symfile directory.

2003-05-04  Klee Dienes <kdienes@apple.com>

	* Makefile: Use GDB_RC_VERSION instead of APPLE_VERSION in the
	gdb.sh target.  Use GDB_FRAMEWORKS and BINUTILS_FRAMEWORKS instead
	of FRAMEWORKS in the install target.

2003-05-04  Klee Dienes <kdienes@apple.com>

	* cache-symfiles.sh: Use the correct path to GDB.  Don't read the
	user's .gdbinit.  Check for dangling symlinks in /usr/lib.

2003-04-24  Jim Ingham  <jingham@apple.com>

	* Makefile: Don't chmod the dylds if you aren't doing 
	--enable-shared.
	Also change the version string to 268.
	
2003-04-25  Klee Dienes  <kdienes@apple.com>


SpamAssassin/clamav/clamav-0.88.2/freshclam/freshclam.c
/*
 *  Copyright (C) 2002 - 2005 Tomasz Kojm <tkojm@clamav.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


#if HAVE_CONFIG_H
#include "clamav-config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <signal.h>
#include <time.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <pwd.h>
#include <grp.h>

#if defined(USE_SYSLOG) && !defined(C_AIX)
#include <syslog.h>
#endif

#include "options.h"
#include "manager.h"
#include "defaults.h"
#include "freshclam.h"
#include "output.h"
#include "target.h"
#include "misc.h"
#include "execute.h"

static short terminate = 0;
extern int active_children;

static void daemon_sighandler(int sig) {

    switch(sig) {
	case SIGCHLD:
	    waitpid(-1, NULL, WNOHANG);
	    active_children--;
	    break;

	case SIGALRM:
	case SIGUSR1:
	    terminate = -1;
	    break;

	case SIGHUP:
	    terminate = -2;
	    break;

	default:
	    terminate = 1;
	    break;
    }

    return;
}


static void writepid(char *pidfile) {
	FILE *fd;
	int old_umask;
    old_umask = umask(0006);
    if((fd = fopen(pidfile, "w")) == NULL) {
	logg("!Can't save PID to file %s: %s\n", pidfile, strerror(errno));
    } else {
	fprintf(fd, "%d", (int) getpid());
	fclose(fd);
    }
    umask(old_umask);
}


int freshclam(struct optstruct *opt)
{
	int ret = 52;
	char *newdir, *cfgfile;
	char *pidfile = NULL;
	struct cfgstruct *copt, *cpt;
	struct sigaction sigact;
	struct sigaction oldact;
#if !defined(C_CYGWIN)  && !defined(C_OS2)
	char *unpuser;
	struct passwd *user;
#endif
	struct stat statbuf;

    if(optc(opt, 'h')) {
	free_opt(opt);
    	help();
    }

    /* parse the config file */
    if((cfgfile = getargl(opt, "config-file"))) {
	copt = parsecfg(cfgfile, 1);
    } else {
	/* TODO: force strict permissions on freshclam.conf */
	if((copt = parsecfg((cfgfile = CONFDIR"/freshclam.conf"), 1)) == NULL)
	    copt = parsecfg((cfgfile = CONFDIR"/clamd.conf"), 1);
    }

    if(!copt) {
	mprintf("!Can't parse the config file %s\n", cfgfile);
	return 56;
    }

    if(optl(opt, "http-proxy") || optl(opt, "proxy-user"))
	mprintf("WARNING: Proxy settings are now only configurable in the config file.\n");

    if(cfgopt(copt, "HTTPProxyPassword")) {
	if(stat(cfgfile, &statbuf) == -1) {
	    mprintf("@Can't stat %s (critical error)\n", cfgfile);
	    return 56;
	}
#ifndef C_CYGWIN
	if(statbuf.st_mode & (S_IRGRP | S_IWGRP | S_IXGRP | S_IROTH | S_IWOTH | S_IXOTH)) {
	    mprintf("@Insecure permissions (for HTTPProxyPassword): %s must have no more than 0700 permissions.\n", cfgfile);
	    return 56;
	}
#endif
    }

#if !defined(C_CYGWIN)  && !defined(C_OS2)
    /* freshclam shouldn't work with root privileges */
    if(optc(opt, 'u')) {
	unpuser = getargc(opt, 'u');
    } else if((cpt = cfgopt(copt, "DatabaseOwner"))) {
	unpuser = cpt->strarg;
    } else {
	unpuser = UNPUSER;
    }

    if(!geteuid()) {
	if((user = getpwnam(unpuser)) == NULL) {
	    mprintf("@Can't get information about user %s.\n", unpuser);
	    exit(60); /* this is critical problem, so we just exit here */
	}

	if(cfgopt(copt, "AllowSupplementaryGroups")) {
#ifdef HAVE_INITGROUPS
	    if(initgroups(unpuser, user->pw_gid)) {
		mprintf("@initgroups() failed.\n");
		exit(61);
	    }
#endif
	} else {
#ifdef HAVE_SETGROUPS
	    if(setgroups(1, &user->pw_gid)) {
		mprintf("@setgroups() failed.\n");
		exit(61);
	    }
#endif
	}

	if(setgid(user->pw_gid)) {
	    mprintf("@setgid(%d) failed.\n", (int) user->pw_gid);
	    exit(61);
	}

	if(setuid(user->pw_uid)) {
	    mprintf("@setuid(%d) failed.\n", (int) user->pw_uid);
	    exit(61);
	}
    }
#endif

    /* initialize some important variables */

    if(optl(opt, "debug") || cfgopt(copt, "Debug"))
	cl_debug();

    if(optc(opt, 'v'))
	mprintf_verbose = 1;

    if(optl(opt, "quiet"))
	mprintf_quiet = 1;

    if(optl(opt, "stdout"))
	mprintf_stdout = 1;

    if(optc(opt, 'V')) {
	print_version();
	exit(0);
    }

    /* initialize logger */

    if(cfgopt(copt, "LogVerbose"))
	logg_verbose = 1;

    if(optc(opt, 'l')) {
	logg_file = getargc(opt, 'l');
	if(logg("--------------------------------------\n")) {
	    mprintf("!Problem with internal logger.\n");
	    exit(62);
	}
    } else if((cpt = cfgopt(copt, "UpdateLogFile"))) {
	logg_file = cpt->strarg; 
	if(logg("--------------------------------------\n")) {
	    mprintf("!Problem with internal logger.\n");
	    exit(62);
	}
    } else
	logg_file = NULL;

#if defined(USE_SYSLOG) && !defined(C_AIX)
    if(cfgopt(copt, "LogSyslog")) {
	    int fac = LOG_LOCAL6;

	if((cpt = cfgopt(copt, "LogFacility"))) {
	    if((fac = logg_facility(cpt->strarg)) == -1) {
		mprintf("!LogFacility: %s: No such facility.\n", cpt->strarg);
		exit(62);
	    }
	}

	openlog("freshclam", LOG_PID, fac);
	logg_syslog = 1;
	syslog(LOG_INFO, "Daemon started.\n");
    }
#endif

    /* change the current working directory */
    if(optl(opt, "datadir")) {
	newdir = getargl(opt, "datadir");
    } else {
	if((cpt = cfgopt(copt, "DatabaseDirectory")))
	    newdir = cpt->strarg;
	else
	    newdir = VIRUSDBDIR;
    }

    if(chdir(newdir)) {
	mprintf("Can't change dir to %s\n", newdir);
	exit(50);
    } else
	mprintf("*Current working dir is %s\n", newdir);


    if(optc(opt, 'd')) {
	    int bigsleep, checks;
	    time_t now, wakeup;

	memset(&sigact, 0, sizeof(struct sigaction));
	sigact.sa_handler = daemon_sighandler;

	if(optc(opt, 'c')) {
	    checks = atoi(getargc(opt, 'c'));
	} else if((cpt = cfgopt(copt, "Checks"))) {
	    checks = cpt->numarg;
	} else {
	    checks = CL_DEFAULT_CHECKS;
	}

	if(checks <= 0) {
	    mprintf("@Number of checks must be a positive integer.\n");
	    exit(41);
	}

	if(!cfgopt(copt, "DNSDatabaseInfo")) {
	    if(checks > 50) {
		mprintf("@Number of checks must be between 1 and 50.\n");
		exit(41);
	    }
	}

	bigsleep = 24 * 3600 / checks;

	if(optc(opt, 'D')) {
	logg("freshclam running in foreground");
	} else {
	if(!cfgopt(copt, "Foreground"))
	    daemonize();
	}

	if (optc(opt, 'p')) {
	    pidfile = getargc(opt, 'p');
	} else if ((cpt = cfgopt(copt, "PidFile"))) {
	    pidfile = cpt->strarg;
	}
	if (pidfile) {
	    writepid(pidfile);
	}

	active_children = 0;

	logg("freshclam daemon "VERSION" (OS: "TARGET_OS_TYPE", ARCH: "TARGET_ARCH_TYPE", CPU: "TARGET_CPU_TYPE")\n");

	sigaction(SIGTERM, &sigact, NULL);
	sigaction(SIGHUP, &sigact, NULL);
	sigaction(SIGINT, &sigact, NULL);
        sigaction(SIGCHLD, &sigact, NULL);

	while(!terminate) {
	    ret = download(copt, opt);

            if(ret > 1) {
		    const char *arg = NULL;

	        if(optl(opt, "on-error-execute"))
		    arg = getargl(opt, "on-error-execute");
		else if((cpt = cfgopt(copt, "OnErrorExecute")))
		    arg = cpt->strarg;

		if(arg)
		    execute("OnErrorExecute", arg);
	    }

	    logg("--------------------------------------\n");
	    sigaction(SIGALRM, &sigact, &oldact);
	    sigaction(SIGUSR1, &sigact, &oldact);
	    time(&wakeup);
	    wakeup += bigsleep;
	    alarm(bigsleep);
	    do {
		pause();
		time(&now);
	    } while (!terminate && now < wakeup);

	    if (terminate == -1) {
		logg("Received signal: wake up\n");
		terminate = 0;
	    } else if (terminate == -2) {
		logg("Received signal: re-opening log file\n");
		terminate = 0;
		logg_close();
	    }

	    sigaction(SIGALRM, &oldact, NULL);
	    sigaction(SIGUSR1, &oldact, NULL);
	}

    } else
	ret = download(copt, opt);

    if(optl(opt, "on-error-execute")) {
	if(ret > 1)
	    system(getargl(opt, "on-error-execute"));

    } else if((cpt = cfgopt(copt, "OnErrorExecute"))) {
	if(ret > 1)
	    system(cpt->strarg);
    }
    if (pidfile) {
        unlink(pidfile);
    }

    return(ret);
}

int download(const struct cfgstruct *copt, const struct optstruct *opt)
{
	int ret = 0, try = 0, maxattempts = 0;
	struct cfgstruct *cpt;


    if((cpt = cfgopt(copt, "MaxAttempts")))
	maxattempts = cpt->numarg;
    else
	maxattempts = CL_DEFAULT_MAXATTEMPTS;


    mprintf("*Max retries == %d\n", maxattempts);

    if((cpt = cfgopt(copt, "DatabaseMirror")) == NULL) {
	mprintf("@You must specify at least one database mirror.\n");
	return 56;
    } else {

	while(cpt) {
	    ret = downloadmanager(copt, opt, cpt->strarg);
	    alarm(0);

	    if(ret == 52 || ret == 54 || ret == 58 || ret == 59) {
		if(try < maxattempts - 1) {
		    mprintf("Trying again in 5 secs...\n");
		    logg("Trying again in 5 secs...\n");
		    try++;
		    sleep(5);
		    continue;
		} else {
		    mprintf("Giving up on %s...\n", cpt->strarg);
		    logg("Giving up on %s...\n", cpt->strarg);
		    cpt = (struct cfgstruct *) cpt->nextarg;
		    if(!cpt) {
			mprintf("@Update failed. Your network may be down or none of the mirrors listed in freshclam.conf is working.\n");
			logg("ERROR: Update failed. Your network may be down or none of the mirrors listed in freshclam.conf is working.\n");
		    }
		    try = 0;
		}

	    } else {
		return ret;
	    }
	}
    }

    return ret;
}

void daemonize(void)
{
	int i;

    for(i = 0; i < 3; i++)
	close(i);

    umask(0);

    if(fork())
	exit(0);

    setsid();
    mprintf_disabled = 1;
}

void help(void)
{
    mprintf_stdout = 1;

    mprintf("\n");
    mprintf("                   Clam AntiVirus: freshclam  "VERSION"\n");
    mprintf("    (C) 2002 - 2005 ClamAV Team - http://www.clamav.net/team.html\n\n");

    mprintf("    --help               -h              show help\n");
    mprintf("    --version            -V              print version number and exit\n");
    mprintf("    --verbose            -v              be verbose\n");
    mprintf("    --debug                              enable debug messages\n");
    mprintf("    --quiet                              only output error messages\n");
    mprintf("    --stdout                             write to stdout instead of stderr\n");
    mprintf("\n");
    mprintf("    --config-file=FILE                   read configuration from FILE.\n");
    mprintf("    --log=FILE           -l FILE         log into FILE\n");
    mprintf("    --daemon             -d              run in daemon mode\n");
    mprintf("    --pid=FILE           -p FILE         save daemon's pid in FILE\n");
    mprintf("    --user=USER          -u USER         run as USER\n");
    mprintf("    --no-dns                             force old non-DNS verification method\n");
    mprintf("    --checks=#n          -c #n           number of checks per day, 1 <= n <= 50\n");
    mprintf("    --datadir=DIRECTORY                  download new databases into DIRECTORY\n");
#ifdef BUILD_CLAMD
    mprintf("    --daemon-notify[=/path/clamd.conf]   send RELOAD command to clamd\n");
#endif
    mprintf("    --local-address=IP   -a IP           bind to IP for HTTP downloads\n");
    mprintf("    --on-update-execute=COMMAND          execute COMMAND after successful update\n");
    mprintf("    --on-error-execute=COMMAND           execute COMMAND if errors occured\n");
    mprintf("    --on-outdated-execute=COMMAND        execute COMMAND when software is outdated\n");
    mprintf("\n");
    exit(0);
}
